src/net/gateway-web3.ts

```diff
@@ -190,69 +195,129 @@ export class Web3Gateway {
+        this._hasTimeOutLastRequest = false
+        return promiseWithTimeout(
+                this.getMetrics(ens), timeout, "The Web3 Gateway is too slow"
+            )
+            .catch((error) => {
+                // TODO refactor errors
+                if (error && error.message == "The Web3 Gateway is too slow") {
+                    this._hasTimeOutLastRequest = true
+                }
+                throw error
+            })
+    }
+
+    /**
+     * The needed metrics for evaluating Gateways during the discovery process are called here.
+     * It also will calculate the response time for each call as a metric itself
+     */
+    private getMetrics(ens: boolean): Promise<void> {
+        let metrics = [
+            this.getBlockNumber(),
+            this.getPeers(),
+            this.isSyncing()
+        ]
+        if (ens) metrics.push(this.checkEns())
+        return Promise.all(metrics).then(() => {
+                // ok
+                // maybe flag it as already checked
+            })
     }
 
     /** Determines whether the current Web3 provider is syncing blocks or not. Several types of prviders may always return false. */
-    public isSyncing(): Promise<boolean> {
-        if (!this._provider) return Promise.resolve(false)
+    public isSyncing(): Promise<void> {
+        if (!this._provider) return Promise.reject()
         else if (this._provider instanceof JsonRpcProvider || this._provider instanceof Web3Provider || this._provider instanceof IpcProvider || this._provider instanceof InfuraProvider) {
-            return this._provider.send("eth_syncing", []).then(result => !!result)
+            let performanceTime = new Date().getTime()
+            return this._provider.send("eth_syncing", []).then(result => {
+                performanceTime = Math.round(new Date().getTime() - performanceTime)
+                this.performanceTime = performanceTime > this.performanceTime ? performanceTime : this.performanceTime
+                return !!result ? Promise.reject() : Promise.resolve()
+            })
         }
-        // else if (this._provider instanceof FallbackProvider || this._provider instanceof EtherscanProvider) {}
 
-        return Promise.resolve(false)
+        return Promise.reject()
     }
 
     /** Request the amount of peers the Gateway is currently connected to */
-    public getPeers(): Promise<number> {
-        if (!this._provider) return Promise.resolve(0)
+    public getPeers(): Promise<void> {
+        // TODO Review if not rejecting and setting peerCount = -1 is the best solution
+        if (!this._provider) return Promise.reject()
         else if (!(this._provider instanceof JsonRpcProvider) && !(this._provider instanceof Web3Provider) &&
             !(this._provider instanceof IpcProvider) && !(this._provider instanceof InfuraProvider)) {
-            return Promise.resolve(0)
+            return Promise.reject()
```

Comment:
There are two changes in this function. The first one is about making the metric fetch "standard" as a `Promise<void>` and the second one is to set the `peerCount` to -1 if the web3 Gateway doesn't support the call, avoiding the usage of `0` as an error and using it as a real number. This is important when selecting the healthy nodes, because if `peerCount = -1` then there is no restriction for rejecting the Gateway, as we know that the call was not supported and therefore we cannot decide based on the number of peers.